File: LeetCode/Easy/0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return construct(nums, 0, nums.length - 1);
    }

    private TreeNode construct(int[] nums, int low, int high) {
        if (low > high) return null;

        int mid = (low + high) / 2;
        
        TreeNode node = new TreeNode(nums[mid]);
        node.left = construct(nums, low, mid - 1);
        node.right = construct(nums, mid + 1, high);

        return node;
    }
}

File: LeetCode/Easy/0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun sortedArrayToBST(nums: IntArray): TreeNode? {
        fun construct(low: Int, high: Int): TreeNode? {
            if (low > high) return null

            val mid = (low + high) / 2
            
            val node = TreeNode(nums[mid])
            node.left = construct(low, mid - 1)
            node.right = construct(mid + 1, high)
            
            return node
        }

        return construct(0, nums.size - 1)
    }
}